// ibmapm-embed/lib/jso.js
// Copyright IBM Corp. 2017. All Rights Reserved.
// Node module: ibmapm
// This file is licensed under the Apache License 2.0.
// License text available at https://opensource.org/licenses/Apache-2.0
'use strict';
var fs = require('fs');
var zlib = require('zlib');
var mm = require('./metric-manager.js').metricManager;
var os = require('os');
var config = require('./config.js');
var adaptor = require('./encode/diagAdaptor');
var reqManager = require('./request-manager.js');
var restClient = require('ibmapm-restclient');

var log4js = require('log4js');
var logger = log4js.getLogger('knj_log');

var cfg = config.getConfig();
var cfgDeepDive = cfg.deepDive;

// request types
function RequestType(id, name) {
    this.id = id;
    this.name = name;
}

function getCurrTime() {
    return (new Date()).getTime();
}

exports.http = new RequestType(1, 'HTTP');
exports.db = new RequestType(21, 'DB');
exports.fs = new RequestType(22, 'FS');
exports.loopback = new RequestType(23, 'LOOPBACK');
exports.mq = new RequestType(24, 'MQ');
exports.others = new RequestType(31, 'UNKNOWN');

var requestTypes = [exports.http, exports.db, exports.fs,
    exports.loopback, exports.mq, exports.others
];
var requestStringToTypes = {
    HTTP: exports.http,
    DB: exports.db,
    FS: exports.fs,
    'HTTP-OUTBOUND': exports.http,
    'HTTP OUTBOUND': exports.http,
    MONGO: exports.db,
    MYSQL: exports.db,
    MQTT: exports.mq,
    MQLIGHT: exports.mq,
    LEVELDOWN: exports.db,
    REDIS: exports.db,
    RIAK: exports.db,
    MEMCACHED: exports.db,
    ORACLEDB: exports.db,
    ORACLE: exports.db,
    'STRONG-ORACLE': exports.db,
    POSTGRES: exports.db,
    LOOPBACK: exports.loopback,
    'LOOPBACK-DATASOURCE-JUGGLER': exports.loopback
};

// JSO structures
function reqDictionaryEntry(id, req) {
    return {
        cid: 10,
        id: id,
        req: req
    };
}

/* function consumerStatus(af) {
    return {
        cid: 1,
        af: af
    };
} */

// {"cid":2,"ci":[
//   {"cid":3,"key":"lurl","vt":"java.lang.String","val":"corbaloc:rir:\/NameServiceServerRoot"},
//   {"cid":3,"key":"requestType","vt":"java.lang.String","val":"JNDI"}]}
function contextDataGroup(ci) {
    return {
        cid: 2,
        ci: ci
    };
}

// {"cid":3,"key":"lurl","vt":"java.lang.String","val":"corbaloc:rir:\/NameServiceServerRoot"}
function contextDataItem(key, val) {
    return {
        cid: 3,
        key: key,
        vt: 'java.lang.String',
        val: val ? (typeof (val) === 'object' ? JSON.stringify(val) : val + '') : ' '
    };
}

// {"cid":4,"props":
// {"num.ear.files":"9",
// "class.path":"C:\\ravi\\programs\\IBM\\Websphere80\\profiles\\dc72002\/properties;..."
// "cpu.speed":"2192","host.name":"ibmwork",
// "install.directory":"C:\\ravi\\programs\\IBM\\Websphere80\\profiles\\dc72002",
// "java.vendor":"IBM Corporation","num.cpus":"8",
// "cell.name":"ibmworkNode04Cell","num.cpus.online":"8",
// "total.memory":"17057431552","os.name":"Windows 7",
// "java.security.policy":"C:\\IBM\\ITM\\dchome\\7.2.0.0.2\\itcamdc\\etc\\datacollector.policy",
// "java.home":"C:\\ravi\\programs\\IBM\\Websphere80\\java\\jre",
// "java.command.line.arguments":
// "C:\\ravi\\programs\\IBM\\Websphere80\/java\/bin\/java -Declipse.security ..."
// "os.version":"6.1","server.name":"server1","num.ejb.modules":"7",
// "java.version":"1.6.0","node.name":"ibmworkNode04",
// "host.address":"ibmwork\/9.76.12.62","java.max.memory":"268435456",
// "num.web.modules":"17","num.jdbc.connection.pools":"6",
// "java.vm.info":"JRE 1.6.0 Windows 7 amd64-64 Compressed References 20110418_80450 ...",
// "was.version":"8.0.0.0",
// "java.vm.name":"IBM J9 VM",
// "transports":"   *:9047 HTTPS,    *:9064 HTTP,    *:9446 HTTPS,    *:9083 HTTP"}}
// props -- key/value pairs
/* function environmentData(props) {
    return {
        cid: 4,
        props: props
    };
} */

// {"cid":5,"id":1,"cn":"ivtServer","mn":"","msig":""}
function methodDictionaryEntry(id, className, methodName, methodSign) {
    return {
        cid: 5,
        id: id,
        cn: className,
        mn: methodName,
        msig: methodSign
    };
}

function contextData(cxtData) {
    if (cxtData) {
        var items = [];
        for (var item in cxtData) {
            if (item !== 'requestHeader' || global.SECURITY_OFF) {
                items.push(contextDataItem(item, cxtData[item]));
            }
        }
        return contextDataGroup(items);
    }
}

// {"cid":6,"time":1315336874037,"type":4,"tid":163,"pid":4,"mid":29,"pmid":31,"rt":0,"cpu":0}
function methodEndInstance(methodEnd) {
    // threadId, time, methId, parentMethId, respTime,
    // cpuTime, stackTrace, cxtData
    return {
        cid: 6,
        time: methodEnd.time,
        type: 4,
        tid: methodEnd.threadId,
        pid: 1,
        mid: methodEnd.methId,
        pmid: methodEnd.parentMethId,
        rt: Math.round(methodEnd.respTime),
        cpu: Math.round(methodEnd.cpuTime),
        st: methodEnd.stackTrace,
        cd: contextData(methodEnd.cxtData)
    };
}

// {"cid":7,"time":1315336873990,"type":3,"tid":163,"pid":4,"mid":30,"pmid":26}
function methodStartInstance(threadId, time, methId, parentMethId) {
    return {
        cid: 7,
        time: time,
        type: 3,
        tid: threadId,
        pid: 1,
        mid: methId,
        pmid: parentMethId
    };
}

// {"cid":8,"time":1315336872833,"id":2,"samp":1,"trt":16,"tcpu":15625000,"maxrt":16,"minrt":16}
function methodSummaryData(ms) {
    return {
        cid: 8,
        id: ms.id,
        samp: ms.samples,
        trt: Math.round(ms.totalResp),
        tcpu: Math.round(ms.totalCpu),
        maxrt: Math.round(ms.maxResp),
        minrt: Math.round(ms.minResp),
        maxct: Math.round(ms.maxCpu),
        minct: Math.round(ms.minCpu)
    };
}

// {"cid":9,"pn":"heap.collection","ver":"1.0","id":2} 9{"cid":9,"pn":"jms",
// "ver":"1.0","id":1,"rpp":["jms.JMS"]}
/* function component(id, comp, version, methProbes, reqProbes) {
    return {
        cid: 9,
        id: id,
        pn: comp,
        ver: version,
        mpp: methProbes,
        rpp: reqProbes
    };
} */

function edgeSeq(edge) {
    if (edge)
        return 1;
    else
        return 2;
}

// {"cid":11,"time":1371606553713,"type":2,"tid":127,"pid":8,"seq":2,"reqtype":2,
// "reqname":"lookup","reqmid":2,"rt":1954,"cpu":390002500,
// "st":"at javax.naming.InitialContext.lookup(InitialContext.java:431)\n......",
// "cd":{"cid":2,"ci":[
// {"cid":3,"key":"lurl","vt":"java.lang.String","val":"corbaloc:rir:\/NameServiceServerRoot"},
// {"cid":3,"key":"requestType","vt":"java.lang.String","val":"JNDI"},
// {"cid":3,"key":"requestName","vt":"java.lang.String","val":"lookup"},
// {"cid":3,"key":"lookupString","vt":"java.lang.String","val":"ejb\/ivtEJBObject"}]}}
function reqEndInstance(reqEnd) {
    // threadId, time, edge, reqType, reqName, reqMethId,
    // respTime, cpuTime, stackTrace, cxtData
    return {
        cid: 11,
        time: reqEnd.time,
        type: 2,
        tid: reqEnd.threadId,
        pid: 1,
        seq: edgeSeq(reqEnd.edge),
        reqtype: reqEnd.reqType,
        reqname: reqEnd.reqName,
        reqmid: reqEnd.reqMethId,
        rt: Math.round(reqEnd.respTime),
        cpu: Math.round(reqEnd.cpuTime),
        st: reqEnd.stackTrace,
        cd: (reqEnd.reqType === 21 && !global.SECURITY_OFF) ?
            contextData([]) : contextData(reqEnd.cxtData),
        requid: reqEnd.reqMethId
    };
}

// {"cid":12,"time":1315336872787,"type":1,"tid":163,"pid":5,"seq":1,"reqtype":1,
// "reqname":"TradeScenarioServlet","reqmid":1}
function reqStartInstance(threadId, time, edge, reqType, reqName, reqMethId) {
    return {
        cid: 12,
        time: time,
        type: 1,
        tid: threadId,
        pid: 1,
        seq: edgeSeq(edge),
        reqtype: reqType,
        reqname: reqName,
        reqmid: reqMethId,
        requid: reqMethId
    };
}

// {"cid":13,"time":1315337141912,"id":1,"samp":1,"trt":63,"tcpu":31250000,"maxrt":63,
// "minrt":63,"seq":1,"missed":0,"basert":63,"basecpu":31250000,"reqname":"_register"}
function reqSummary(rs, reqName) {
    return {
        cid: 13,
        id: rs.typeId,
        time: rs.startTime,
        seq: edgeSeq(rs.edge),
        samp: rs.samples,
        trt: Math.round(rs.totalResp),
        tcpu: Math.round(rs.totalCpu),
        maxrt: Math.round(rs.maxResp),
        minrt: Math.round(rs.minResp),
        missed: 0,
        basert: Math.round(rs.totalResp / rs.samples),
        basecpu: Math.round(rs.totalCpu / rs.samples),
        reqname: reqName
    };
}

// {"cid":14,"tid":127,"tname":"WebContainer : 0","ttype":"webContainer","pn":"WAS"}
/* function threadDictEntry() {
    return {
        cid: 14,
        id: 1,
        tname: 'Event Loop',
        ttype: 'Event Loop',
        pn: 'node.js'
    };
} */

// {"cid":17,"ver":7200,"minclient":7201}
function version() {
    return {
        cid: 17,
        ver: 7200,
        minclient: 7201
    };
}

// {"cid":18,"name":"WAS","ver":"1.0","act":true,
// "profnames":["ejb","jca","jdbc","jndi","servlet","jms","webServices"],
// "refnames":["ejb","jca","jdbc","jndi","servlet","jms","webServices"]}
function product(name, version, active, profiles, refs) {
    return {
        cid: 18,
        name: name,
        ver: version,
        act: active,
        profnames: profiles,
        refs: refs
    };
}

var debugPort = 3000;
var hostname = os.hostname().split('.').shift();
var port;
var dir;
var fileNamePrefix = 'dfe_data_' + process.pid + '_';

// JSO file handling configuration
// - eventsPerFile -- when this number of events is reached then JSO file will be committed
//                    once all top level requests are finished
// - fileCommitTime -- when this file is not modified for more than specified number of
//                     seconds then it will be committed
// - maxFiles -- when number of JSO files in directory reaches this value then the oldest
//               one is deleted
// When the file is committed then all subsequent writes to it are ignored
// (i.e. for not yet finished requests).

function JsoFile(path, name) {
    this.events = 0;
    this.activeRequests = 0;
    this.committed = false;
    this.buffers = [];
    this.totalLength = 0;
    this.samplingCount = 0;
    this.startTime = 0;
    this.endTime = 0;
    this.methSummary = {};
    this.nextMethId = 1;

    this.path = path;
    this.name = name;

    // write header
    this.write(version());
    this.write(product('node.js', '1.0', true, [], []));
    var self = this;
    requestTypes.forEach(function(type) {
        self.write(reqDictionaryEntry(type.id, type.name));
    });

    this.reqSummary = {};
}

var writeBufferToREST = function(b, jso) {
    var appid = hostname + '_' + port;
    logger.debug('writing buffer to REST: ' + b);
    logger.info('Sending deepdive data for ', appid, ' from ', jso.startTime, ' to ', jso.endTime);
    zlib.gzip(b, function(err, result) {
        if (err) {
            logger.error(err);
            return;
        }
        var wrappeddata;
        var theIpaddr = appid;
        if (process.env.HYBRID_BMAPPID && process.env.HYBRID_BMAPPID !== 'undefined') {
            theIpaddr = process.env.HYBRID_BMAPPID;
        }

        wrappeddata = adaptor.addEncoder(result, {
            ipaddr: theIpaddr,
            start: jso.startTime,
            end: jso.endTime
        });

        restClient.sendJSO(wrappeddata);
    });
};


JsoFile.prototype.read = function(filepath) {
    var b = fs.readFileSync(filepath);
    return b;
};

JsoFile.prototype.write_code = function(s) {
    var txt = JSON.stringify(s);
    logger.debug('writing', txt);
    var len = Buffer.byteLength(txt, 'utf8');
    var b = new Buffer(2 + len);
    b.writeInt16BE(len, 0);
    b.write(txt, 2, undefined, 'utf8');
    this.buffers.push(b);
    this.totalLength += b.length;
};

JsoFile.prototype.write = function(s, sync) {
    this.write_code(s);
};

// write summary, close and rename to jso
JsoFile.prototype.commit = function() {
    var self = this;

    if (!self.committed) {
        self.endTime = getCurrTime();
        // write request summary
        for (var rn in self.reqSummary) {
            if (self.reqSummary.hasOwnProperty(rn)){
                var rs = self.reqSummary[rn];
                logger.debug('writing reqSummary', rn, 'samples', rs.samples);
                if (rs.samples > 0) {
                    self.write(reqSummary(rs, rn), true);
                }
            }
        }

        // write method dictionary and summary
        for (var mn in self.methSummary) {
            if (self.methSummary.hasOwnProperty(mn)){
                var ms = self.methSummary[mn];
                self.write(methodDictionaryEntry(ms.id, '', mn, ''), true);
                if (ms.samples > 0) {
                    self.write(
                        methodSummaryData(ms),
                        true
                    );
                }
            }
        }

        var contents = Buffer.concat(self.buffers, self.totalLength);
        try {
            writeBufferToREST(contents, self);
        } catch (e) {
            logger.error('Failed to send jso to REST service: ', e);
            logger.debug('Failed to send jso to REST service: ', e.stack);
        }
        if (process.env.KNJ_LOG_LEVEL === 'all') {
            // Dump the jso file for the debug mode on BAM.
            fs.open(self.path + self.name + '.jso', 'w', function(err, fd) {
                if (err) {
                    logger.error('Failed to open jso file: ' + self.path + self.name + '.jso');
                    logger.error(err);
                } else {
                    var contents = Buffer.concat(self.buffers, self.totalLength);
                    fs.write(fd, contents, 0, contents.length, null);
                    fs.close(fd);
                    removeOldFiles();
                }
            });
        }
        self.committed = true;
        reqManager.resetEgdeReqCount();
    }
};

JsoFile.prototype.startRequest = function(threadId, reqId, timer, edge, type, name) {
    var self = this;
    type = requestStringToTypes[type];
    if (!type) {
        type = exports.others;
    }
    if (edge) {
        cfgDeepDive = config.getConfig().deepDive;
    }

    /* switch to sampling on threadId which is actually the request Id */
    if (!self.committed && !(self.samplingCount % cfgDeepDive.sampling) &&
        (cfgDeepDive.methodTrace || edge)) {
        if (edge) {
            self.activeRequests += 1;
            self.events += 1;
        }

        if (!(name in self.reqSummary)) {
            self.reqSummary[name] = {
                typeId: type.id,
                startTime: timer.startTimeMillis,
                edge: edge,
                samples: 0,
                totalResp: 0,
                totalCpu: 0
            };
            logger.debug('added request to reqSummary', name);
        } else {
            logger.debug('request was already in reqSummary', name);
        }
        self.write(
            reqStartInstance(threadId, timer.startTimeMillis, edge, type.id, name, reqId),
            true
        );
    }
};

JsoFile.prototype.stopRequest = function(threadId, reqId, timer, edge,
    type, name, stackTrace, cxtData) {
    var self = this;
    type = requestStringToTypes[type];
    if (!type) {
        type = exports.others;
    }

    /* switch to sampling on threadId which is actually the request Id */
    if (!self.committed && !(self.samplingCount % cfgDeepDive.sampling) &&
        (cfgDeepDive.methodTrace || edge)) {
        if (edge) {
            --self.activeRequests;
        }

        var rs = self.reqSummary[name];
        if (rs) {
            logger.debug('updating request', name);

            rs.totalResp += timer.timeDelta;
            rs.totalCpu += timer.cpuTimeDelta;
            if (rs.samples === 0 || rs.maxResp < timer.timeDelta) {
                rs.maxResp = timer.timeDelta;
            }
            if (rs.samples === 0 || rs.minResp > timer.timeDelta) {
                rs.minResp = timer.timeDelta;
            }
            rs.samples += 1;
            self.write(
                reqEndInstance({
                    threadId: threadId,
                    time: timer.startTimeMillis + Math.floor(timer.timeDelta),
                    edge: edge,
                    reqType: type.id,
                    reqName: name,
                    reqMethId: reqId,
                    respTime: timer.timeDelta,
                    cpuTime: timer.cpuTimeDelta,
                    stackTrace: stackTrace,
                    cxtData: cxtData
                }),
                true
            );
        } else {
            logger.debug('request not found in summary', name);
        }

        if (self.activeRequests === 0 && !self.committed &&
            self.events >= cfgDeepDive.eventsPerFile) {
            logger.info('committing by max events reached', self.events, cfgDeepDive.eventsPerFile);
            self.commit();
        }
    }
};

JsoFile.prototype.startMethod = function(threadId, timer, name, parentId) {
    var self = this;
    /* switch to sampling on threadId which is actually the request Id */
    if (!self.committed && cfgDeepDive.methodTrace &&
        !(self.samplingCount % cfgDeepDive.sampling)) {

        var methId;

        var ms = self.methSummary[name];
        if (ms) {
            methId = ms.id;
        } else {
            methId = parseInt((timer.startTimeMillis % 100000).toString() + self.nextMethId);
            self.nextMethId += 1;
            self.methSummary[name] = {
                id: methId,
                samples: 0,
                totalResp: 0,
                totalCpu: 0
            };
        }

        self.write(methodStartInstance(threadId, timer.startTimeMillis, methId, parentId), true);

        return methId;
    }
};

JsoFile.prototype.stopMethod = function(threadId, timer, name, parentId, stackTrace, cxtData) {

    var self = this;

    /* switch to sampling on threadId which is actually the request Id */
    if (!self.committed && cfgDeepDive.methodTrace &&
        !(self.samplingCount % cfgDeepDive.sampling)) {

        var ms = self.methSummary[name];
        if (ms) {
            ms.totalResp += timer.timeDelta;
            ms.totalCpu += timer.cpuTimeDelta;
            if (ms.samples === 0 || ms.maxResp < timer.timeDelta) {
                ms.maxResp = timer.timeDelta;
            }
            if (ms.samples === 0 || ms.minResp > timer.timeDelta) {
                ms.minResp = timer.timeDelta;
            }
            ms.samples += 1;
            self.write(
                methodEndInstance({
                    threadId: threadId,
                    time: timer.startTimeMillis + Math.floor(timer.timeDelta),
                    methId: ms.id,
                    parentMethId: parentId,
                    respTime: timer.timeDelta,
                    cpuTime: timer.cpuTimeDelta,
                    stackTrace: stackTrace,
                    cxtData: cxtData
                }),
                true
            );
        }
    }
};

// open

var currentJso;
var nextJsoId = 1;

var emptyJso = {
    write: function() {
        // emptyJso
    },

    commit: function() { reqManager.resetEgdeReqCount(); },

    startRequest: function() {
        // emptyJso
    },

    stopRequest: function() {
        // emptyJso
    },

    startMethod: function() {
        // emptyJso
    },

    stopMethod: function() {
        // emptyJso
    }
};

function initDirAndPort() {
    port = (typeof v8debug === 'object') ? debugPort : mm.metric.port;
    port = port || 'N/A';
    dir = './dchome/';
}

function mkdir(dir) {
    var s = dir.split('/');
    var d = '';
    if (dir.indexOf('.') === 0) {
        s.forEach(function(e) {
            d = d + e + '/';
            if (!fs.existsSync(d)) {
                fs.mkdirSync(d);
            }
        });
    } else {
        s.forEach(function(e) {
            d = d + '/' + e;
            if (!fs.existsSync(d)) {
                fs.mkdirSync(d);
            }
        });
    }
}

function removeOldFiles() {
    var maxFiles = cfgDeepDive.maxFiles;
    try {
        fs.readdir(dir, function(err, files) {
            if (err) {
                logger.error('Failed to read dir:', dir);
                logger.error(err);
            }
            var r = new RegExp(dir + fileNamePrefix + '(\\d*)_(\\d*).jso');
            files = files.map(function(name) {
                return dir + name;
            }).filter(function(name) {
                var matchResult = r.test(name);
                if (!matchResult && name.indexOf('.tmp') === -1) {
                    fs.unlink(name);
                }
                return matchResult;
            });

            if (files.length <= maxFiles) return;

            var nums = files.map(function(name) {
                return {
                    time: r.exec(name)[1],
                    index: parseInt(r.exec(name)[2])
                };
            });
            nums = nums.sort(function(a, b) {
                return a.index - b.index;
            }).slice(0, nums.length - maxFiles);
            nums.forEach(function(num) {
                fs.unlink(dir + fileNamePrefix + num.time + '_' + num.index + '.jso');
            });
        });
    } catch (e) {
        logger.error('removeOldFiles failed: ', e);
    }
}

exports.open = function() {
    cfgDeepDive = config.getConfig().deepDive;
    initDirAndPort();
    if (currentJso) {
        if (currentJso.committed) {
            currentJso = null;
        } else {
            return currentJso;
        }
    }
    if (port || process.env.KNJ_LOG_LEVEL === 'all') {
        mkdir(dir);
        var name = fileNamePrefix + (new Date()).getTime() + '_' + nextJsoId;
        logger.debug('opening file', dir, '/', name);
        nextJsoId += 1;
        var jso = new JsoFile(dir, name);
        jso.startTime = getCurrTime();
        currentJso = jso;
        const timeoutObj = setTimeout(function() {
            logger.info('Commiting file by timeout', jso.name);
            jso.commit();
        }, cfgDeepDive.fileCommitTime * 1000);
        timeoutObj.unref();
        return currentJso;
    } else {
        logger.info('Unable to store data into jso file, as port is not initalized.');
        return emptyJso;
    }
};
